qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -225,68 +205,77 @@ def run(self, dag):
         # Replace source instructions with target translations.
 
         replace_start_time = time.time()
-        for node in dag.op_nodes():
-            node_qargs = tuple(qarg_indices[bit] for bit in node.qargs)
-            qubit_set = frozenset(node_qargs)
 
-            if node.name in target_basis:
-                continue
-            if (
-                node_qargs in self._qargs_with_non_global_operation
-                and node.name in self._qargs_with_non_global_operation[node_qargs]
-            ):
-                continue
+        def apply_translation(dag):
+            for node in dag.op_nodes():
+                node_qargs = tuple(qarg_indices[bit] for bit in node.qargs)
+                qubit_set = frozenset(node_qargs)
+                if node.name in target_basis:
+                    if isinstance(node.op, ControlFlowOp):
+                        flow_blocks = []
+                        for block in node.op.blocks:
+                            dag_block = circuit_to_dag(block)
+                            flow_dag_block = apply_translation(dag_block)
+                            flow_circ_block = dag_to_circuit(flow_dag_block)
+                            flow_blocks.append(flow_circ_block)
+                        node.op = node.op.replace_blocks(flow_blocks)
+                    continue
+                if (
+                    node_qargs in self._qargs_with_non_global_operation
+                    and node.name in self._qargs_with_non_global_operation[node_qargs]
+                ):
+                    continue
 
-            if dag.has_calibration_for(node):
-                continue
+                if dag.has_calibration_for(node):
+                    continue
 
-            def replace_node(node, instr_map):
-                target_params, target_dag = instr_map[node.op.name, node.op.num_qubits]
-                if len(node.op.params) != len(target_params):
-                    raise TranspilerError(
-                        "Translation num_params not equal to op num_params."
-                        "Op: {} {} Translation: {}\n{}".format(
-                            node.op.params, node.op.name, target_params, target_dag
+                def replace_node(node, instr_map):
+                    target_params, target_dag = instr_map[node.op.name, node.op.num_qubits]
+                    if len(node.op.params) != len(target_params):
```

Comment:
This function is now sufficiently nested (and its outer one sufficiently long) that it's probably worth factoring this out into its own standalone function.  On a quick scan through, I think the function only closes over `dag`, so we can probably just add that as its first parameter, and move the function definition out to be a stand-alone private module function.

Reply:
done in 8e35e93d7